fineract-provider/src/main/java/org/apache/fineract/infrastructure/springbatch/messagehandler/JmsBatchWorkerMessageListener.java

```diff
@@ -0,0 +1,97 @@
+import org.apache.fineract.infrastructure.springbatch.messagehandler.conditions.JmsWorkerCondition;
+import org.springframework.batch.core.BatchStatus;
+import org.springframework.batch.core.Step;
+import org.springframework.batch.core.StepExecution;
+import org.springframework.batch.core.explore.JobExplorer;
+import org.springframework.batch.core.repository.JobRepository;
+import org.springframework.batch.core.step.StepLocator;
+import org.springframework.batch.integration.partition.StepExecutionRequest;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.jms.support.converter.MessagingMessageConverter;
+import org.springframework.messaging.Message;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+@Conditional(JmsWorkerCondition.class)
+public class JmsBatchWorkerMessageListener implements MessageListener, InitializingBean {
+
+    private final JobRepository jobRepository;
+    private final StepLocator stepLocator;
+    private final JobExplorer jobExplorer;
+    private final InputChannelInterceptor inputInterceptor;
+    private MessagingMessageConverter converter;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        converter = new MessagingMessageConverter();
+    }
+
+    public void handleMessage(StepExecutionRequest request) {
+        Long jobExecutionId = request.getJobExecutionId();
+        Long stepExecutionId = request.getStepExecutionId();
+        String stepName = request.getStepName();
+
+        StepExecution stepExecution = jobExplorer.getStepExecution(jobExecutionId, stepExecutionId);
+        if (stepExecution == null) {
+            throw new IllegalStateException("stepExecution cannot be null");
+        }
+
+        Step step = stepLocator.getStep(stepName);
+        try {
+            step.execute(stepExecution);
+            stepExecution.setStatus(BatchStatus.COMPLETED);
+        } catch (Exception e) {
+            stepExecution.addFailureException(e);
+            stepExecution.setStatus(BatchStatus.FAILED);
+        } finally {
+            jobRepository.update(stepExecution);
+        }
+    }
+
+    @Override
+    @SuppressWarnings({ "unchecked" })
+    public void onMessage(javax.jms.Message message) {
+        try {
+            Message<ContextualMessage> msg = (Message<ContextualMessage>) converter.fromMessage(message);
+            Message<StepExecutionRequest> requestMessage = inputInterceptor.beforeHandleMessage(msg);
+            handleMessage(requestMessage.getPayload());
+        } catch (Exception e) {
+            log.error("Exception while processing JMS message", e);
+        } finally {
+            try {
+                message.acknowledge();
+            } catch (JMSException e) {
+                log.error("Unable to acknowledge message", e);
+            }
+        }
+    }
+}
```

Comment:
Arnold few questions :
1.We setting step execution status from worker in job repository. Is this a full step or a step with partition only which the worker was executing?
2.How and when does the partition for which the worker failed gets picked up by other worker. During the same job execution or a new execution? What is desired behavior?
3.Should acknowledgement be sent even if the handling of message fails? Do we need to configure ACK timeout?

Reply:
@ruchiD:
1. Only the partition
2. The partition messages are put onto Queues and not topic. This guarantees one thing, only a single consumer will receive a message at once. Since client acknowledgement is used, only one worker will process the message at once. If that worker does not ack the message because it crashed (i.e. the connection to the broker is gone) it'll be redelivered to another worker.
3. Not sure what you mean by ACK timeout, can you please explain?

Reply:
Timeout for acknowledgement of message for redelivery. but not required in this scenario.